From 5480eb36dcd3073a896a1d90a2b4934a177bf7b0 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Wed, 10 Jan 2024 17:35:44 +0100
Subject: [PATCH 09/19] net: phy: add support for a driver-defined software
 reset function

Some ethernet PHYs don't fully reset from BMCR_RESET and require custom
mdio message.

Add a soft_reset function to struct phy_driver and call it in phy_init
when available.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 drivers/net/phy/phy.c | 3 +++
 include/phy.h         | 4 ++++
 2 files changed, 7 insertions(+)

diff --git a/drivers/net/phy/phy.c b/drivers/net/phy/phy.c
index 63b3e46f101..374bce86d2d 100644
--- a/drivers/net/phy/phy.c
+++ b/drivers/net/phy/phy.c
@@ -716,6 +716,9 @@ int phy_reset(struct phy_device *phydev)
 	if (phydev->flags & PHY_FLAG_BROKEN_RESET)
 		return 0;
 
+	if (phydev->drv->soft_reset)
+		return phydev->drv->soft_reset(phydev);
+
 #ifdef CONFIG_PHYLIB_10G
 	/* If it's 10G, we need to issue reset through one of the MMDs */
 	if (phydev->is_c45) {
diff --git a/include/phy.h b/include/phy.h
index ae23814bbf3..902163b8410 100644
--- a/include/phy.h
+++ b/include/phy.h
@@ -103,6 +103,10 @@ struct phy_driver {
 
 	u32 features;
 
+	/* perform driver defined software reset */
+	/* called from phy_reset */
+	int (*soft_reset)(struct phy_device *phydev);
+
 	/* Called to do any driver startup necessities */
 	/* Will be called during phy_connect */
 	int (*probe)(struct phy_device *phydev);
-- 
2.43.0

